lms/djangoapps/instructor/tests/test_api.py

```diff
@@ -2022,6 +2033,190 @@ def test_get_students_features_cohorted(self, is_cohorted):
+        )
+        invoice_transaction.save()
+        course_registration_code = CourseRegistrationCode(
+            code='abcde',
+            course_id=self.course.id.to_deprecated_string(),
+            created_by=self.instructor,
+            invoice=self.sale_invoice_1,
+            invoice_item=self.invoice_item,
+            mode_slug='honor'
+        )
+        course_registration_code.save()
+
+        admin_user = AdminFactory()
+        admin_cart = Order.get_cart_for_user(admin_user)
+        PaidCourseRegistration.add_to_order(admin_cart, self.course.id)
+        admin_cart.purchase()
+
+        test_user = UserFactory()
+        redeem_url = reverse('register_code_redemption', args=[course_registration_code.code])
+        self.client.login(username=test_user.username, password='test')
+        response = self.client.get(redeem_url)
+        self.assertEquals(response.status_code, 200)
+        # check button text
+        self.assertTrue('Activate Course Enrollment' in response.content)
+
+        response = self.client.post(redeem_url)
+        self.assertEquals(response.status_code, 200)
+
+        CourseFinanceAdminRole(self.course.id).add_users(self.instructor)
+        user_profile_1 = UserProfileFactory.create(user=self.students[0])
+        user_profile_1.meta = '{"company": "asdasda"}'
+        user_profile_1.save()
+
+        # UserFactory.profile(self.students[0], create=True, extracted=None, {"meta": {"company": 'Test com'}})
+        self.client.login(username=self.instructor.username, password='test')
+        url = reverse('get_enrollment_report', kwargs={'course_id': self.course.id.to_deprecated_string()})
+        response = self.client.get(url, {})
+        self.assertIn('Your detailed enrollment report is being generated!', response.content)
+
+    def test_bulk_purchase_detailed_report(self):
+        """
+        test to generate detailed enrollment report.
+        purchase registration codes and activate course
+        enrollment using purchased registration code and
+        then generate the enrollment report from the
+        instructor dashboard to check if the report is generated
+        successfully.
+        """
+        paid_course_reg_item = PaidCourseRegistration.add_to_order(self.cart, self.course.id)
+        # update the quantity of the cart item paid_course_reg_item
+        resp = self.client.post(reverse('shoppingcart.views.update_user_cart'),
+                                {'ItemId': paid_course_reg_item.id, 'qty': '4'})
+        self.assertEqual(resp.status_code, 200)
+        # apply the coupon code to the item in the cart
+        resp = self.client.post(reverse('shoppingcart.views.use_code'), {'code': self.coupon_code})
+        self.assertEqual(resp.status_code, 200)
+        self.cart.purchase()
+
+        course_reg_codes = CourseRegistrationCode.objects.filter(order=self.cart)
+        redeem_url = reverse('register_code_redemption', args=[course_reg_codes[0].code])
```

Comment:
This block of code that registers via redeem code seems to be copied a number of times. This is a prime candidate for refactoring into a helper method: `register_with_redemption_code(user, code)`.